fix(logger): prevent overwriting standard keys

dreamorosi opened the pull request:
## Summary

### Changes

> Please provide a summary of what's being changed

<!-- What is this PR solving? Write a clear description or reference the issue(s) it addresses. -->

This PR fixes a bug that allowed customer payloads to override Powertools-managed keys in logs.

Logger has a set of keys that are added to every log entry, for example `level`, `timestamp`, `service`, and more. These keys are what make logs discoverable.

Due to a bug, in some cases when customers appended keys to the Logger instance or passed extra keys to a log method, some of the service keys would be overridden by the customer-provided keys.

For example:

```ts
const user = {
  name: "john",
  level: "assistant"
};

logger.debug("user info", user);
```

In this case the `level` key from the user object would override the level in the log entry.

This PR fixes this bug and makes it so that when a customer-provided key conflicts with a reserved key, Logger will drop the customer one and log a warning.

> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** fixes #3217

<!-------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/CONTRIBUTING.md#sending-a-pull-request
- Check that there isn't already a PR that addresses the same issue. If you find a duplicate, please leave a comment under the existing PR so we can discuss how to move forward
- Check that the change meets the project's tenets https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets
- Add a PR title that follows the conventional commit semantics - https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2
- If relevant, add tests that prove that the change is effective and works
- Whenever relevant, make sure to comment functions/methods/types and make appropriate changes to the documentation
------->

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

dreamorosi commented:
Ran e2e tests and they're [passing](https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/13136167149).

leandrodamascena commented:
I'll review this today morning.

leandrodamascena reviewed (changes requested):
Hi @dreamorosi, I left a comment about protected keys.

Another thing is that even I consider this a bug because the customer may end up having some logs with `level: 'BLABLA` and this is not a valid log level + we should enforce good practices. But this is a breaking change both in terms of possible data loss and in the customers' CI with warning popping up, and we should highlight this in the documentation/release notes and monitor if we have customers complaining.

My vote is to move forward with this PR, but monitor the feedback after releasing it.

dreamorosi commented on packages/logger/src/constants.ts:
```diff
@@ -37,4 +37,18 @@ const LogLevelThreshold = {
+ *
+ * These keys are reserved and cannot be overwritten by custom log attributes.
+ */
+const ReservedKeys = [
+  'level',
+  'message',
+  'sampling_rate',
+  'service',
+  'timestamp',
+  'xray_trace_id',
+];
```
I got the list from our docs [here](https://docs.powertools.aws.dev/lambda/typescript/latest/core/logger/#standard-structured-keys).

Error is not included because customers are always supposed to pass it, doing this for example

```ts
logger.debug({
  message: "a less important error happened",
  error,
})
```

Will emit a debug log with the error formatted at the `error` key.

For the `xray_trace_id` I'm ambivalent, the logger already reads it from the environment so customers don't have to pass it, but I don't know if this is a thing outside of Lambda so I'm ok removing it for now.

I'll make a commit to remove it.

dreamorosi commented:
> Another thing is that even I consider this a bug because the customer may end up having some logs with `level: 'BLABLA` and this is not a valid log level + we should enforce good practices. But this is a breaking change both in terms of possible data loss and in the customers' CI with `warnings` popping up in their pipeline and possible breaking this. We must highlight this in the documentation/release notes and monitor if we have customers complaining.
> 
> My vote is to move forward with this PR, but monitor the feedback after releasing it.

I added two callouts to the docs, one in the structured keys section:

<img width="1120" alt="image" src="https://github.com/user-attachments/assets/7b1c1796-f994-4933-bdc9-30d32945ff7c" />

and one in the section that talks about adding keys (the link goes to the previously mentioned one):

<img width="902" alt="image" src="https://github.com/user-attachments/assets/2d123ffd-79e0-4a53-a08f-b42e33c52137" />

I also updated the API docs for the two methods so they include the same info:

see in IDE:

![image](https://github.com/user-attachments/assets/27f78e87-7f49-4da1-b9ea-0a8b303f89bf)

and in the API docs:

![image](https://github.com/user-attachments/assets/e2416f42-134a-4a03-b82f-6be8ce0c1f9c)

---

I also improved the types of the methods so that customers can get an error in their IDE if they try to set one of these keys.
